src/FakeItEasy/Creation/DummyValueResolver.cs

```diff
@@ -26,57 +26,86 @@ public DummyValueResolver(DynamicDummyFactory dummyFactory, IFakeObjectCreator f
             this.strategies = new ResolveStrategy[]
                 {
                     new ResolveFromDummyFactoryStrategy(dummyFactory),
-                    new ResolveByCreatingTaskStrategy(this),
-                    new ResolveByCreatingLazyStrategy(this),
+                    new ResolveByCreatingTaskStrategy(),
+                    new ResolveByCreatingLazyStrategy(),
                     new ResolveByActivatingValueTypeStrategy(),
-                    new ResolveByCreatingFakeStrategy(fakeObjectCreator, this),
-                    new ResolveByInstantiatingClassUsingDummyValuesAsConstructorArgumentsStrategy(this)
+                    new ResolveByCreatingFakeStrategy(fakeObjectCreator),
+                    new ResolveByInstantiatingClassUsingDummyValuesAsConstructorArgumentsStrategy()
                 };
         }
 
-        public CreationResult TryResolveDummyValue(DummyCreationSession session, Type typeOfDummy)
+        public CreationResult TryResolveDummyValue(Type typeOfDummy)
         {
-            if (!session.TryBeginToResolveType(typeOfDummy))
-            {
-                return CreationResult.FailedToCreateDummy(typeOfDummy, "Recursive dependency detected. Already resolving " + typeOfDummy + '.');
-            }
-
-            var creationResult = this.TryResolveDummyValueWithAllAvailableStrategies(session, typeOfDummy);
-            if (creationResult.WasSuccessful)
+            if (this.strategyCache.TryGetValue(typeOfDummy, out ResolveStrategy cachedStrategy))
             {
-                session.OnSuccessfulResolve(typeOfDummy);
+                return cachedStrategy.TryCreateDummyValue(typeOfDummy, this);
             }
 
-            return creationResult;
+            var resolver = new LoopDetectingResovler(this.strategies, this.strategyCache);
+            return resolver.TryResolveDummyValueWithAllAvailableStrategies(typeOfDummy);
         }
 
-        private CreationResult TryResolveDummyValueWithAllAvailableStrategies(DummyCreationSession session, Type typeOfDummy)
+        private class LoopDetectingResovler : IDummyValueResolver
         {
-            if (this.strategyCache.TryGetValue(typeOfDummy, out ResolveStrategy cachedStrategy))
+            private readonly ResolveStrategy[] strategies;
+            private readonly ConcurrentDictionary<Type, ResolveStrategy> strategyCache;
+            private readonly HashSet<Type> typesCurrentlyBeingResolved;
+
+            public LoopDetectingResovler(ResolveStrategy[] strategies, ConcurrentDictionary<Type, ResolveStrategy> strategyCache)
             {
-                return cachedStrategy.TryCreateDummyValue(session, typeOfDummy);
+                this.strategies = strategies;
+                this.strategyCache = strategyCache;
+                this.typesCurrentlyBeingResolved = new HashSet<Type>();

```

Comment:
Isn't creating this immediately basically the same as creating the `DummyCreationSession` eagerly?
Maybe it could be created lazily; not sure how much of a benefit that would be.

Reply:
Ah, no. In many many cases (I hope), we'll be creating a Dummy of a type we've attempted to use before, so the `DummyValueResolver` will find a strategy in its cache and create the Dummy without ever making a `LoopDetectingResolver`.
However, when there's no cached strategy, we create `LoopDetectingResolver` and the `DummyValueResolver` calls `TryResolveDummyValueWithAllAvailableStrategies` immediately, so `typesCurrentlyBeingResolved` will be used.

Reply:
Ah, I see